{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "22527e43",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ea9d214e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= pd.read_csv(\"dmel-all-r6.31.gtf\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3763fba1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>19961297</td>\n",
       "      <td>19969323</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0031081\"; gene_symbol \"Nep3\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>19961689</td>\n",
       "      <td>19968479</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>5UTR</td>\n",
       "      <td>19961689</td>\n",
       "      <td>19961845</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>19961689</td>\n",
       "      <td>19961845</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X</td>\n",
       "      <td>FlyBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>19963955</td>\n",
       "      <td>19964071</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; tra...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0        1     2         3         4  5  6  7  \\\n",
       "0  X  FlyBase  gene  19961297  19969323  .  +  .   \n",
       "1  X  FlyBase  mRNA  19961689  19968479  .  +  .   \n",
       "2  X  FlyBase  5UTR  19961689  19961845  .  +  .   \n",
       "3  X  FlyBase  exon  19961689  19961845  .  +  .   \n",
       "4  X  FlyBase  exon  19963955  19964071  .  +  .   \n",
       "\n",
       "                                                   8  \n",
       "0         gene_id \"FBgn0031081\"; gene_symbol \"Nep3\";  \n",
       "1  gene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; tra...  \n",
       "2  gene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; tra...  \n",
       "3  gene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; tra...  \n",
       "4  gene_id \"FBgn0031081\"; gene_symbol \"Nep3\"; tra...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "de55de8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns=['seqname','source','feature','start','end','Xscore','strand','Xframe','attributes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0a0bd3e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[['gene id','1','2','3','4','5']]= df.attributes.str.split(\";\",expand=True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d39a2be5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop(df[df.feature != 'gene'].index, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "45a50bb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['iso1']=df['end']-df['start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dc802890",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= df.drop(['start','seqname','source','attributes','feature','strand','end','Xscore','Xframe','1','2','3','4','5'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e2d43f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= df.sort_values(by=['iso1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e15180ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene id</th>\n",
       "      <th>iso1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>307664</th>\n",
       "      <td>gene_id \"FBgn0085765\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435196</th>\n",
       "      <td>gene_id \"FBgn0267503\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435242</th>\n",
       "      <td>gene_id \"FBgn0267518\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435260</th>\n",
       "      <td>gene_id \"FBgn0267524\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435175</th>\n",
       "      <td>gene_id \"FBgn0267496\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      gene id  iso1\n",
       "307664  gene_id \"FBgn0085765\"    29\n",
       "435196  gene_id \"FBgn0267503\"    29\n",
       "435242  gene_id \"FBgn0267518\"    29\n",
       "435260  gene_id \"FBgn0267524\"    29\n",
       "435175  gene_id \"FBgn0267496\"    29"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "88993d7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df.to_csv('copy_number.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7193b8ee",
   "metadata": {},
   "source": [
    "# This is where the first \"script\" would be made, and stop. Once the ref csv was started, the below script would be run with all the files in the working directory "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "103f0b64",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4616a968",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= pd.read_csv(\"copy_number.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c4969f29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_id \"FBgn0085765\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_id \"FBgn0267518\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_id \"FBgn0267524\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_id \"FBgn0267510\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_id \"FBgn0267503\"</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       0  length\n",
       "0  gene_id \"FBgn0085765\"      29\n",
       "1  gene_id \"FBgn0267518\"      29\n",
       "2  gene_id \"FBgn0267524\"      29\n",
       "3  gene_id \"FBgn0267510\"      29\n",
       "4  gene_id \"FBgn0267503\"      29"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "7d7c899c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1= pd.read_csv(\"B59_liftoff\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ce25019e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.columns=['seqname','source','feature','start','end','Xscore','strand','Xframe','attributes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9ca9bc66",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.drop(df1[df1.feature != 'gene'].index, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "5c440f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1[['gene id','1','2','3','4','5','6','7','8']]= df1.attributes.str.split(\";\",expand=True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0448dec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1['sample 2']=df1['end']-df1['start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2011cc73",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1= df1.drop(['start','seqname','source','attributes','feature','strand','end','Xscore','Xframe','1','2','3','4','5','6','7','8'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "625ac1e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1= df1.sort_values(by=['sample 2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e8f72307",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Xscore</th>\n",
       "      <th>strand</th>\n",
       "      <th>Xframe</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene id</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>gene</td>\n",
       "      <td>89123</td>\n",
       "      <td>91075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0031208\"; gene_symbol \"CR11023\"; ...</td>\n",
       "      <td>gene_id \"FBgn0031208\"</td>\n",
       "      <td>gene_symbol \"CR11023\"</td>\n",
       "      <td>coverage \"0.998\"</td>\n",
       "      <td>sequence_ID \"0.996\"</td>\n",
       "      <td>extra_copy_number \"0\"</td>\n",
       "      <td>copy_num_ID \"FBgn0031208_0\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>89123</td>\n",
       "      <td>89709</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0031208\"; gene_symbol \"CR11023\"; ...</td>\n",
       "      <td>gene_id \"FBgn0031208\"</td>\n",
       "      <td>gene_symbol \"CR11023\"</td>\n",
       "      <td>transcript_id \"FBtr0475186\"</td>\n",
       "      <td>transcript_symbol \"CR11023-RE\"</td>\n",
       "      <td>Parent \"FBgn0031208\"</td>\n",
       "      <td>extra_copy_number \"0\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>89786</td>\n",
       "      <td>91075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0031208\"; gene_symbol \"CR11023\"; ...</td>\n",
       "      <td>gene_id \"FBgn0031208\"</td>\n",
       "      <td>gene_symbol \"CR11023\"</td>\n",
       "      <td>transcript_id \"FBtr0475186\"</td>\n",
       "      <td>transcript_symbol \"CR11023-RE\"</td>\n",
       "      <td>Parent \"FBgn0031208\"</td>\n",
       "      <td>extra_copy_number \"0\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>89123</td>\n",
       "      <td>91075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0031208\"; gene_symbol \"CR11023\"; ...</td>\n",
       "      <td>gene_id \"FBgn0031208\"</td>\n",
       "      <td>gene_symbol \"CR11023\"</td>\n",
       "      <td>transcript_id \"FBtr0475186\"</td>\n",
       "      <td>transcript_symbol \"CR11023-RE\"</td>\n",
       "      <td>Parent \"FBgn0031208\"</td>\n",
       "      <td>extra_copy_number \"0\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>gene</td>\n",
       "      <td>91430</td>\n",
       "      <td>102955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0002121\"; gene_symbol \"l(2)gl\"; c...</td>\n",
       "      <td>gene_id \"FBgn0002121\"</td>\n",
       "      <td>gene_symbol \"l(2)gl\"</td>\n",
       "      <td>coverage \"0.996\"</td>\n",
       "      <td>sequence_ID \"0.993\"</td>\n",
       "      <td>extra_copy_number \"0\"</td>\n",
       "      <td>copy_num_ID \"FBgn0002121_0\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname   source     feature  start     end Xscore strand Xframe  \\\n",
       "0  Chr-2L  Liftoff        gene  89123   91075      .      +      .   \n",
       "1  Chr-2L  Liftoff        exon  89123   89709      .      +      .   \n",
       "2  Chr-2L  Liftoff        exon  89786   91075      .      +      .   \n",
       "3  Chr-2L  Liftoff  pseudogene  89123   91075      .      +      .   \n",
       "4  Chr-2L  Liftoff        gene  91430  102955      .      -      .   \n",
       "\n",
       "                                          attributes                gene id  \\\n",
       "0  gene_id \"FBgn0031208\"; gene_symbol \"CR11023\"; ...  gene_id \"FBgn0031208\"   \n",
       "1  gene_id \"FBgn0031208\"; gene_symbol \"CR11023\"; ...  gene_id \"FBgn0031208\"   \n",
       "2  gene_id \"FBgn0031208\"; gene_symbol \"CR11023\"; ...  gene_id \"FBgn0031208\"   \n",
       "3  gene_id \"FBgn0031208\"; gene_symbol \"CR11023\"; ...  gene_id \"FBgn0031208\"   \n",
       "4  gene_id \"FBgn0002121\"; gene_symbol \"l(2)gl\"; c...  gene_id \"FBgn0002121\"   \n",
       "\n",
       "                        1                             2  \\\n",
       "0   gene_symbol \"CR11023\"              coverage \"0.998\"   \n",
       "1   gene_symbol \"CR11023\"   transcript_id \"FBtr0475186\"   \n",
       "2   gene_symbol \"CR11023\"   transcript_id \"FBtr0475186\"   \n",
       "3   gene_symbol \"CR11023\"   transcript_id \"FBtr0475186\"   \n",
       "4    gene_symbol \"l(2)gl\"              coverage \"0.996\"   \n",
       "\n",
       "                                 3                       4  \\\n",
       "0              sequence_ID \"0.996\"   extra_copy_number \"0\"   \n",
       "1   transcript_symbol \"CR11023-RE\"    Parent \"FBgn0031208\"   \n",
       "2   transcript_symbol \"CR11023-RE\"    Parent \"FBgn0031208\"   \n",
       "3   transcript_symbol \"CR11023-RE\"    Parent \"FBgn0031208\"   \n",
       "4              sequence_ID \"0.993\"   extra_copy_number \"0\"   \n",
       "\n",
       "                              5  6     7     8  \n",
       "0   copy_num_ID \"FBgn0031208_0\"     None  None  \n",
       "1         extra_copy_number \"0\"     None  None  \n",
       "2         extra_copy_number \"0\"     None  None  \n",
       "3         extra_copy_number \"0\"     None  None  \n",
       "4   copy_num_ID \"FBgn0002121_0\"     None  None  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f108bba2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.merge(df, df1, how=\"outer\", on=\"gene id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e23f7a63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     gene id     iso1  sample 2\n",
      "0      gene_id \"FBgn0085765\"       29       NaN\n",
      "1      gene_id \"FBgn0267503\"       29       NaN\n",
      "2      gene_id \"FBgn0267518\"       29       NaN\n",
      "3      gene_id \"FBgn0267524\"       29       NaN\n",
      "4      gene_id \"FBgn0267496\"       29       NaN\n",
      "...                      ...      ...       ...\n",
      "17841  gene_id \"FBgn0028369\"   394148  375378.0\n",
      "17842  gene_id \"FBgn0263780\"   396035       NaN\n",
      "17843  gene_id \"FBgn0267429\"   434040       NaN\n",
      "17844  gene_id \"FBgn0267430\"   709420       NaN\n",
      "17845  gene_id \"FBgn0267431\"  1965856       NaN\n",
      "\n",
      "[17846 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "print(df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "76736dcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "unmapped=df2[df2['gene id'] == 'NaN']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "257ac294",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene id</th>\n",
       "      <th>iso1</th>\n",
       "      <th>sample 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene id, iso1, sample 2]\n",
       "Index: []"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unmapped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "23ea7d44",
   "metadata": {},
   "outputs": [],
   "source": [
    "unmapped= df1[~df1['4'].str.contains('0')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "731e8b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "unmapped=df1[~df1['5'].str.contains('0')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "741a305c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Xscore</th>\n",
       "      <th>strand</th>\n",
       "      <th>Xframe</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene id</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43577</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>8625110</td>\n",
       "      <td>8625182</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0051895\"; gene_symbol \"tRNA:Phe-G...</td>\n",
       "      <td>gene_id \"FBgn0051895\"</td>\n",
       "      <td>gene_symbol \"tRNA:Phe-GAA-1-3\"</td>\n",
       "      <td>transcript_id \"FBtr0079692\"</td>\n",
       "      <td>transcript_symbol \"tRNA:Phe-GAA-1-3-RA\"</td>\n",
       "      <td>Parent \"FBgn0051895\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43578</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>8625110</td>\n",
       "      <td>8625182</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0051895\"; gene_symbol \"tRNA:Phe-G...</td>\n",
       "      <td>gene_id \"FBgn0051895\"</td>\n",
       "      <td>gene_symbol \"tRNA:Phe-GAA-1-3\"</td>\n",
       "      <td>transcript_id \"FBtr0079692\"</td>\n",
       "      <td>transcript_symbol \"tRNA:Phe-GAA-1-3-RA\"</td>\n",
       "      <td>Parent \"FBgn0051895\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74412</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>16834138</td>\n",
       "      <td>16834265</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0261068\"; gene_symbol \"LSm7\"; tra...</td>\n",
       "      <td>gene_id \"FBgn0261068\"</td>\n",
       "      <td>gene_symbol \"LSm7\"</td>\n",
       "      <td>transcript_id \"FBtr0343770\"</td>\n",
       "      <td>transcript_symbol \"LSm7-RC\"</td>\n",
       "      <td>Parent \"FBgn0261068\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74413</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>16834160</td>\n",
       "      <td>16834809</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0261068\"; gene_symbol \"LSm7\"; tra...</td>\n",
       "      <td>gene_id \"FBgn0261068\"</td>\n",
       "      <td>gene_symbol \"LSm7\"</td>\n",
       "      <td>transcript_id \"FBtr0080934\"</td>\n",
       "      <td>transcript_symbol \"LSm7-RA\"</td>\n",
       "      <td>Parent \"FBgn0261068\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74414</th>\n",
       "      <td>Chr-2L</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>16834399</td>\n",
       "      <td>16834809</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0261068\"; gene_symbol \"LSm7\"; tra...</td>\n",
       "      <td>gene_id \"FBgn0261068\"</td>\n",
       "      <td>gene_symbol \"LSm7\"</td>\n",
       "      <td>transcript_id \"FBtr0308804\"</td>\n",
       "      <td>transcript_symbol \"LSm7-RB\"</td>\n",
       "      <td>Parent \"FBgn0261068\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198612</th>\n",
       "      <td>Chr-2R</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>20861389</td>\n",
       "      <td>20861410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0263546\"; gene_symbol \"mir-4985\";...</td>\n",
       "      <td>gene_id \"FBgn0263546\"</td>\n",
       "      <td>gene_symbol \"mir-4985\"</td>\n",
       "      <td>transcript_id \"FBtr0309661\"</td>\n",
       "      <td>transcript_symbol \"mir-4985-RA\"</td>\n",
       "      <td>Parent \"FBgn0263546\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198613</th>\n",
       "      <td>Chr-2R</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>20861425</td>\n",
       "      <td>20861446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0263546\"; gene_symbol \"mir-4985\";...</td>\n",
       "      <td>gene_id \"FBgn0263546\"</td>\n",
       "      <td>gene_symbol \"mir-4985\"</td>\n",
       "      <td>transcript_id \"FBtr0309662\"</td>\n",
       "      <td>transcript_symbol \"mir-4985-RB\"</td>\n",
       "      <td>Parent \"FBgn0263546\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198614</th>\n",
       "      <td>Chr-2R</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>pre_miRNA</td>\n",
       "      <td>20861361</td>\n",
       "      <td>20861461</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0263546\"; gene_symbol \"mir-4985\";...</td>\n",
       "      <td>gene_id \"FBgn0263546\"</td>\n",
       "      <td>gene_symbol \"mir-4985\"</td>\n",
       "      <td>transcript_id \"FBtr0309660\"</td>\n",
       "      <td>transcript_symbol \"mir-4985-RM\"</td>\n",
       "      <td>Parent \"FBgn0263546\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524118</th>\n",
       "      <td>Chr-X</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>19488673</td>\n",
       "      <td>19489401</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0265705\"; gene_symbol \"asRNA:CR44...</td>\n",
       "      <td>gene_id \"FBgn0265705\"</td>\n",
       "      <td>gene_symbol \"asRNA:CR44512\"</td>\n",
       "      <td>transcript_id \"FBtr0340613\"</td>\n",
       "      <td>transcript_symbol \"asRNA:CR44512-RA\"</td>\n",
       "      <td>Parent \"FBgn0265705\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524119</th>\n",
       "      <td>Chr-X</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>19488673</td>\n",
       "      <td>19489401</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"FBgn0265705\"; gene_symbol \"asRNA:CR44...</td>\n",
       "      <td>gene_id \"FBgn0265705\"</td>\n",
       "      <td>gene_symbol \"asRNA:CR44512\"</td>\n",
       "      <td>transcript_id \"FBtr0340613\"</td>\n",
       "      <td>transcript_symbol \"asRNA:CR44512-RA\"</td>\n",
       "      <td>Parent \"FBgn0265705\"</td>\n",
       "      <td>extra_copy_number \"1\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname   source    feature     start       end Xscore strand Xframe  \\\n",
       "43577   Chr-2L  Liftoff       exon   8625110   8625182      .      +      .   \n",
       "43578   Chr-2L  Liftoff       tRNA   8625110   8625182      .      +      .   \n",
       "74412   Chr-2L  Liftoff       exon  16834138  16834265      .      -      .   \n",
       "74413   Chr-2L  Liftoff       exon  16834160  16834809      .      -      .   \n",
       "74414   Chr-2L  Liftoff       exon  16834399  16834809      .      -      .   \n",
       "...        ...      ...        ...       ...       ...    ...    ...    ...   \n",
       "198612  Chr-2R  Liftoff      miRNA  20861389  20861410      .      +      .   \n",
       "198613  Chr-2R  Liftoff      miRNA  20861425  20861446      .      +      .   \n",
       "198614  Chr-2R  Liftoff  pre_miRNA  20861361  20861461      .      +      .   \n",
       "524118   Chr-X  Liftoff       exon  19488673  19489401      .      -      .   \n",
       "524119   Chr-X  Liftoff      ncRNA  19488673  19489401      .      -      .   \n",
       "\n",
       "                                               attributes  \\\n",
       "43577   gene_id \"FBgn0051895\"; gene_symbol \"tRNA:Phe-G...   \n",
       "43578   gene_id \"FBgn0051895\"; gene_symbol \"tRNA:Phe-G...   \n",
       "74412   gene_id \"FBgn0261068\"; gene_symbol \"LSm7\"; tra...   \n",
       "74413   gene_id \"FBgn0261068\"; gene_symbol \"LSm7\"; tra...   \n",
       "74414   gene_id \"FBgn0261068\"; gene_symbol \"LSm7\"; tra...   \n",
       "...                                                   ...   \n",
       "198612  gene_id \"FBgn0263546\"; gene_symbol \"mir-4985\";...   \n",
       "198613  gene_id \"FBgn0263546\"; gene_symbol \"mir-4985\";...   \n",
       "198614  gene_id \"FBgn0263546\"; gene_symbol \"mir-4985\";...   \n",
       "524118  gene_id \"FBgn0265705\"; gene_symbol \"asRNA:CR44...   \n",
       "524119  gene_id \"FBgn0265705\"; gene_symbol \"asRNA:CR44...   \n",
       "\n",
       "                      gene id                                1  \\\n",
       "43577   gene_id \"FBgn0051895\"   gene_symbol \"tRNA:Phe-GAA-1-3\"   \n",
       "43578   gene_id \"FBgn0051895\"   gene_symbol \"tRNA:Phe-GAA-1-3\"   \n",
       "74412   gene_id \"FBgn0261068\"               gene_symbol \"LSm7\"   \n",
       "74413   gene_id \"FBgn0261068\"               gene_symbol \"LSm7\"   \n",
       "74414   gene_id \"FBgn0261068\"               gene_symbol \"LSm7\"   \n",
       "...                       ...                              ...   \n",
       "198612  gene_id \"FBgn0263546\"           gene_symbol \"mir-4985\"   \n",
       "198613  gene_id \"FBgn0263546\"           gene_symbol \"mir-4985\"   \n",
       "198614  gene_id \"FBgn0263546\"           gene_symbol \"mir-4985\"   \n",
       "524118  gene_id \"FBgn0265705\"      gene_symbol \"asRNA:CR44512\"   \n",
       "524119  gene_id \"FBgn0265705\"      gene_symbol \"asRNA:CR44512\"   \n",
       "\n",
       "                                   2  \\\n",
       "43577    transcript_id \"FBtr0079692\"   \n",
       "43578    transcript_id \"FBtr0079692\"   \n",
       "74412    transcript_id \"FBtr0343770\"   \n",
       "74413    transcript_id \"FBtr0080934\"   \n",
       "74414    transcript_id \"FBtr0308804\"   \n",
       "...                              ...   \n",
       "198612   transcript_id \"FBtr0309661\"   \n",
       "198613   transcript_id \"FBtr0309662\"   \n",
       "198614   transcript_id \"FBtr0309660\"   \n",
       "524118   transcript_id \"FBtr0340613\"   \n",
       "524119   transcript_id \"FBtr0340613\"   \n",
       "\n",
       "                                               3                      4  \\\n",
       "43577    transcript_symbol \"tRNA:Phe-GAA-1-3-RA\"   Parent \"FBgn0051895\"   \n",
       "43578    transcript_symbol \"tRNA:Phe-GAA-1-3-RA\"   Parent \"FBgn0051895\"   \n",
       "74412                transcript_symbol \"LSm7-RC\"   Parent \"FBgn0261068\"   \n",
       "74413                transcript_symbol \"LSm7-RA\"   Parent \"FBgn0261068\"   \n",
       "74414                transcript_symbol \"LSm7-RB\"   Parent \"FBgn0261068\"   \n",
       "...                                          ...                    ...   \n",
       "198612           transcript_symbol \"mir-4985-RA\"   Parent \"FBgn0263546\"   \n",
       "198613           transcript_symbol \"mir-4985-RB\"   Parent \"FBgn0263546\"   \n",
       "198614           transcript_symbol \"mir-4985-RM\"   Parent \"FBgn0263546\"   \n",
       "524118      transcript_symbol \"asRNA:CR44512-RA\"   Parent \"FBgn0265705\"   \n",
       "524119      transcript_symbol \"asRNA:CR44512-RA\"   Parent \"FBgn0265705\"   \n",
       "\n",
       "                             5  6     7     8  \n",
       "43577    extra_copy_number \"1\"     None  None  \n",
       "43578    extra_copy_number \"1\"     None  None  \n",
       "74412    extra_copy_number \"1\"     None  None  \n",
       "74413    extra_copy_number \"1\"     None  None  \n",
       "74414    extra_copy_number \"1\"     None  None  \n",
       "...                        ... ..   ...   ...  \n",
       "198612   extra_copy_number \"1\"     None  None  \n",
       "198613   extra_copy_number \"1\"     None  None  \n",
       "198614   extra_copy_number \"1\"     None  None  \n",
       "524118   extra_copy_number \"1\"     None  None  \n",
       "524119   extra_copy_number \"1\"     None  None  \n",
       "\n",
       "[70 rows x 18 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04dac2d7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
